### PR TITLE
engine: Add Env to exec.go

### PR DESCRIPTION
**Pull request by adamsigal**
Add functionality to specify environment variables in exec.go.

## Tips:

* please read the style guide before submitting your patch:
[docs/style-guide.md](../docs/style-guide.md)

* commit message titles must be in the form:

```topic: Capitalized message with no trailing period```

or:

```topic, topic2: Capitalized message with no trailing period```

* golang code must be formatted according to the standard, please run:

```
make gofmt		# formats the entire project correctly
```

or format a single golang file correctly:

```
gofmt -w yourcode.go
```

* please rebase your patch against current git master:

```
git checkout master
git pull origin master
git checkout your-feature
git rebase master
git push your-remote your-feature
hub pull-request	# or submit with the github web ui
```

* after a patch review, please ping @purpleidea so we know to re-review:

```
# make changes based on reviews...
git add -p		# add new changes
git commit --amend	# combine with existing commit
git push your-remote your-feature -f
# now ping @purpleidea in the github PR since it doesn't notify us automatically
```

### CODE REVIEW

File: engine/resources/exec.go [hunk 1]
```diff
@@ -138,9 +141,38 @@ func (obj *ExecRes) Validate() error {
...
+	for key := range obj.Env {
+		if !isNameValid(key) {
+			return fmt.Errorf("environment variable name must only contain alphanumeric characters or underscore")
+		}
+	}
+
 	return nil
 }
 
+// TODO: move this function further down in code
```
Comment (purpleidea) [inline engine/resources/exec.go]: Why have a TODO in here?

File: engine/resources/exec.go [hunk 2]
```diff
@@ -138,9 +141,38 @@ func (obj *ExecRes) Validate() error {
...
+// checks that environment variable name is valid
+func isNameValid(varName string) bool {
```
Comment (purpleidea) [inline engine/resources/exec.go]: change sig to return `error` instead. Instead of `return false` use fmt.Errorf(...reason...) instead.

File: engine/resources/exec.go [hunk 3]
```diff
@@ -138,9 +141,38 @@ func (obj *ExecRes) Validate() error {
 		}
 	}
 
+	// check that environment variables' format is valid
+	for key := range obj.Env {
+		if !isNameValid(key) {
+			return fmt.Errorf("environment variable name must only contain alphanumeric characters or underscore")
```
Comment (purpleidea) [inline engine/resources/exec.go]: errwrap.Wrapf please with shorter message now that below sig changes

File: engine/resources/exec.go [hunk 4]
```diff
@@ -138,9 +141,38 @@ func (obj *ExecRes) Validate() error {
...
 	return nil
 }
 
+// TODO: move this function further down in code
+// checks that environment variable name is valid
+func isNameValid(varName string) bool {
+	if word == "" {
+		return false
+	}
+	var i int
```
Comment (purpleidea) [inline engine/resources/exec.go]: not needed. `i := 0` instead.

File: engine/resources/exec.go [hunk 5]
```diff
@@ -138,9 +141,38 @@ func (obj *ExecRes) Validate() error {
...
+	for i = 0; i < len(varName); i++ {
+		c := varName[i]
+		// if first character is number, name is invalid
+		if i == 0 && '0' <= c && c <= '9' {
+			return false
+		}
+		// if name does not only contain alphanumeric characters or underscore,
```
Comment (purpleidea) [inline engine/resources/exec.go]: one line mini comments or full sentences. Full sentences are capitalized and end with a period.

File: engine/resources/exec.go [hunk 6]
```diff
@@ -277,6 +309,16 @@ func (obj *ExecRes) CheckApply(apply bool) (bool, error) {
 		}
 		cmd := exec.Command(cmdName, cmdArgs...)
 		cmd.Dir = obj.IfCwd // run program in pwd if ""
+
+		//TODO: verify with James
```
Comment (purpleidea) [inline engine/resources/exec.go]: Don't put a comment like this =D Ask in review instead ;)

File: engine/resources/exec.go [hunk 7]
```diff
@@ -277,6 +309,16 @@ func (obj *ExecRes) CheckApply(apply bool) (bool, error) {
...
+		cmdEnv := make([]string, len(obj.Env))
```
Comment (purpleidea) [inline engine/resources/exec.go]: I dislike this syntax, instead just do: `cmdEnv := []string{}`

File: engine/resources/exec.go [hunk 8]
```diff
@@ -277,6 +309,16 @@ func (obj *ExecRes) CheckApply(apply bool) (bool, error) {
...
+		i := 0
+		for key := range obj.Env {
```
Comment (purpleidea) [inline engine/resources/exec.go]: Look up the map iteration syntax, don't use a counter like `i`. And we'll want to make this deterministic somehow... I'll let you figure it out. Do you know why it isn't deterministic and why it should be?

File: engine/resources/exec.go (hunk 1 above)
Comment (adamsigal) [reply engine/resources/exec.go]: I thought that this function was out of place in its current position, but I wanted it to be easily visible the first time you were looking at it, since it is related to the neighbouring function.

File: engine/resources/exec.go (hunk 8 above)
Comment (adamsigal) [reply engine/resources/exec.go]: I know that the order of map traversal isn't deterministic in this case. Because of that the end result isn't deterministic because duplicate definitions of a variable aren't assured to be in the same order, possibly leading to an incorrect final definition.
I see the problem, I'll find a solution.

File: engine/resources/exec.go (hunk 3 above)
Comment (adamsigal) [reply engine/resources/exec.go]: I don't quite understand what you mean by "now that below sig changes"

Comment (adamsigal) [reply engine/resources/exec.go]: Sorry, I get it now

File: engine/resources/exec.go (hunk 1 above)
Comment (purpleidea) [reply engine/resources/exec.go]: This doesn't matter, I'd see it anyways because `git diff` shows me the changes. But if a future reader of the code (not the patch) would find it easier if they were co-located, then it makes sense to do so. This depends on convention and other things. Eg: do you layout chained functions in standard execution order or dependency order? Or do private util functions go at the end so they don't interleave with the struct methods? (That is what I do.)

File: engine/resources/exec.go (hunk 8 above)
Comment (purpleidea) [reply engine/resources/exec.go]: Wrong, a map (in golang or mcl) can't have duplicate keys, so _that's_ not the problem. But the list we build will be in different orders which can make debugging or other things tricky. IOW a heisenbug! We never want to cause those. (Eg: "i ran the code and it did X, now it did Y... what's going on? *brain explodes*)

**Comment by purpleidea** [comment]
Ping when ready for a new review. Thanks!

File: engine/resources/exec.go (hunk 8 above)
Comment (adamsigal) [reply engine/resources/exec.go]: So we have to make sure that all of the variables are processed in the order they were written, right? For example we couldn't just sort them alphabetically and process them in that order.

Comment (adamsigal) [reply engine/resources/exec.go]: I can't think of how we would do that in order just by using a map. I'm assuming that a map is still the best option because you confirmed that to me at one point, but I fail to think of a non-hacky solution

Comment (purpleidea) [reply engine/resources/exec.go]: > For example we couldn't just sort them alphabetically and process them in that order.

Why not? Of course we can. The solution is to first pull all the keys out of the map into a new list. Then sort that list. Then loop through that list (now deterministic) and read the values out from the map in that order. TADA =D I'll let you write the code for this. Max 6 lines.

Comment (adamsigal) [reply engine/resources/exec.go]: Okay, I believe I've implemented the necessary changes for this part correctly. I'll submit a new PR in a second and let you take a look.

File: engine/resources/exec.go [hunk 9]
```diff
@@ -138,6 +142,10 @@ func (obj *ExecRes) Validate() error {
 		}
 	}
 
+	// check that environment variables' format is valid
+	for key := range obj.Env {
+		isNameValid(key)
```
Comment (purpleidea) [inline engine/resources/exec.go]: Missing something here...

File: engine/resources/exec.go [hunk 10]
```diff
@@ -277,6 +285,14 @@ func (obj *ExecRes) CheckApply(apply bool) (bool, error) {
 		}
 		cmd := exec.Command(cmdName, cmdArgs...)
 		cmd.Dir = obj.IfCwd // run program in pwd if ""
+
+		cmdEnv := []string{}
+		for key := range obj.Env {
+			cmdEnv = append(cmdEnv, key+"="+obj.Env[key])
```
Comment (purpleidea) [inline engine/resources/exec.go]: This actually _does_ do what we want, but it's non-standard and ends up sorting more than we need to. Try and implement the above described solution that *just* sorts the map keys. They're unique (it's a map) so we don't need to sort the values alongside them.

Comment (adamsigal) [reply engine/resources/exec.go]: Let me make sure I'm understanding correctly. What you're suggesting is
1. Make a list of the keys by iterating through map
2. Sort that list
3. Iterate through sorted list, in each iteration append "<key>=<value>" to cmd.Env

If this is what you mean, isn't it a worse solution to have to use 2 for loops to accomplish the same thing?

File: engine/resources/exec.go (hunk 9 above)
Comment (adamsigal) [reply engine/resources/exec.go]: return, right?

Comment (purpleidea) [reply engine/resources/exec.go]: errwrap.Wrapf...

File: engine/resources/exec.go (hunk 10 above)
Comment (purpleidea) [reply engine/resources/exec.go]: Yes and no. Technically there's a loop inside the Sort() that you're using, and you're sorting additional stuff which isn't needed.

Comment (purpleidea) [reply engine/resources/exec.go]: Plus it would be good to get you to learn the standard sort approach when the map keys aren't sortable or are a different type.

File: engine/resources/exec.go (hunk 9 above)
Comment (adamsigal) [reply engine/resources/exec.go]: Is it not redundant to have error messages in the `isNameValid` function _and_ in the errwrap?

Comment (purpleidea) [reply engine/resources/exec.go]: this currently does nothing

Comment (adamsigal) [reply engine/resources/exec.go]: I changed it in the most recent pull request to 
```
for key := range obj.Env {
	return errwrap.Wrapf(isNameValid(key), "invalid variable name")
}
return nil
```

**Comment by purpleidea** [comment]
If you rebase to current master, the builds should be fixed now.

File: engine/resources/exec.go [hunk 11]
```diff
@@ -54,7 +55,7 @@ type ExecRes struct {
 	// only be used when a Shell is *not* specified. The advantage of this
 	// is that you don't have to worry about escape characters.
 	Args []string `yaml:"args"`
-	// Cmd is the dir to run the command in. If empty, then this will use
+	// Cwd is the dir to run the command in. If empty, then this will use
```
Comment (purpleidea) [inline engine/resources/exec.go]: This is a good catch, but technically it's unrelated to this commit and so it should be in a separate commit. If the commit as-is was perfect except for this, I'd probably make an exception and just merge, but it's good practice for you to learn how to "split a commit" (google or ddg that phrase) so let's do it here! (And remember to fix commit message!)

File: engine/resources/exec.go [hunk 12]
```diff
@@ -277,6 +285,7 @@ func (obj *ExecRes) CheckApply(apply bool) (bool, error) {
 		}
 		cmd := exec.Command(cmdName, cmdArgs...)
 		cmd.Dir = obj.IfCwd // run program in pwd if ""
+
```
Comment (purpleidea) [inline engine/resources/exec.go]: Random nl add? This sort of mistake should be caught by you running `git diff` *before* you submit the patch. Please do =D

File: engine/resources/exec.go [hunk 13]
```diff
@@ -368,6 +377,18 @@ func (obj *ExecRes) CheckApply(apply bool) (bool, error) {
...
 	cmd := exec.CommandContext(ctx, cmdName, cmdArgs...)
 	cmd.Dir = obj.Cwd // run program in pwd if ""
+
+	envKeys := []string{}
+	for i := range obj.Env {
+		envKeys = append(envKeys, i)
+	}
+	sort.Strings(envKeys)
+	cmdEnv := []string{}
+	for k := range envKeys {
```
Comment (purpleidea) [inline engine/resources/exec.go]: Use the `for _, x := range` form and code will look nicer.

File: engine/resources/exec.go [hunk 14]
```diff
@@ -800,3 +821,21 @@ func (obj *wrapWriter) Write(p []byte) (int, error) {
 func (obj *wrapWriter) String() string {
 	return obj.Buffer.String()
 }
+
+// checks that environment variable name is valid
```
Comment (purpleidea) [inline engine/resources/exec.go]: docstring in wrong format for two reasons. Please fix and tell me what the two mistakes are.

File: engine/resources/exec.go [hunk 15]
```diff
@@ -800,3 +821,21 @@ func (obj *wrapWriter) Write(p []byte) (int, error) {
...
+func isNameValid(varName string) error {
+	if varName == "" {
+		return fmt.Errorf("variable name cannot be an empty string")
+	}
+	i := 0
```
Comment (purpleidea) [inline engine/resources/exec.go]: This line and the next can be combined. Know how?

File: engine/resources/exec.go [hunk 16]
```diff
@@ -800,3 +821,21 @@ func (obj *wrapWriter) Write(p []byte) (int, error) {
...
+	for i = 0; i < len(varName); i++ {
+		c := varName[i]
+		if i == 0 && '0' <= c && c <= '9' {
```
Comment (purpleidea) [inline engine/resources/exec.go]: This is yucky. Just use strings.HasPrefix outside the loop.

File: engine/resources/exec.go [hunk 17]
```diff
@@ -800,3 +821,21 @@ func (obj *wrapWriter) Write(p []byte) (int, error) {
...
+			return fmt.Errorf("variable name cannot begin with number")
+		}
+		if !(c == '_' || '0' <= c && c <= '9' || 'a' <= c && c <= 'z' || 'A' <= c && c <= 'Z') {
```
Comment (purpleidea) [inline engine/resources/exec.go]: O.K., but double check there isn't a more succinct or idiomatic alternative in the `strings` package.

File: engine/resources/resources_test.go [hunk 18]
```diff
@@ -286,6 +286,44 @@ func TestResources1(t *testing.T) {
...
+		r := makeRes("exec", "x1")
+		res := r.(*ExecRes) // if this panics, the test will panic
+		s := "hello, world"
+		f := "/tmp/whatever"
+		res.Cmd = fmt.Sprintf("echo '%s' > '%s'", s, f)
+		res.Shell = "/bin/bash"
+		res.IfCmd = "! diff <(cat /tmp/whatever) <(echo hello, world)"
+		res.IfShell = "/bin/bash"
+		res.WatchCmd = fmt.Sprintf("/usr/bin/inotifywait -e modify -m %s", f)
+		res.Env = map[string]string{
```
Comment (purpleidea) [inline engine/resources/resources_test.go]: It would be logical to put this higher next to the part of the resource it goes with, right?

File: engine/resources/resources_test.go [hunk 19]
```diff
@@ -286,6 +286,44 @@ func TestResources1(t *testing.T) {
...
+			"boiling": "one hundred", "freezing": "zero", "room": "twenty-five",
```
Comment (purpleidea) [inline engine/resources/resources_test.go]: I like =D
But put one on each line instead of the awkward split of three and then two.

File: engine/resources/resources_test.go [hunk 20]
```diff
@@ -286,6 +286,44 @@ func TestResources1(t *testing.T) {
...
+			"house": "twenty-two", "canada": "negative thirty",
+		}
+		//res.WatchShell = "/bin/bash"
+
+		timeline := []Step{
```
Comment (purpleidea) [inline engine/resources/resources_test.go]: Do we ever check in here somewhere that `Env` actually did anything? IOW if you took all the `Env` code in CheckApply and other methods away, would this test still pass? If so, then it's missing something.

File: examples/lang/exec_env.mcl [hunk 21]
```diff
@@ -0,0 +1,7 @@
+exec "exec2" {
+	shell => "/bin/bash",
+	ifcmd => "/bin/true",
+	ifshell => "/bin/bash",
+	env => {"BOILING" => "one hundred",},
+    cmd => "/usr/bin/env",
```
Comment (purpleidea) [inline examples/lang/exec_env.mcl]: wat! (tabs!)

File: engine/resources/exec.go (hunk 13 above)
Comment (adamsigal) [reply engine/resources/exec.go]: do you want me to do this for all of my for loops?

File: engine/resources/exec.go (hunk 14 above)
Comment (adamsigal) [reply engine/resources/exec.go]: No trailing period, and the comment doesn't begin with the function name.

File: engine/resources/exec.go (hunk 15 above)
Comment (adamsigal) [reply engine/resources/exec.go]: I believe lines 830 and 831 can be combined like so:
```
for i := 0; i < len(varName); i++ {
```

File: engine/resources/exec.go (hunk 16 above)
Comment (adamsigal) [reply engine/resources/exec.go]: Maybe I'm missing something, but wouldn't I have to loop through characters 0-9 if I used strings.HasPrefix? That doesn't sound like the simplest solution.

Comment (adamsigal) [reply engine/resources/exec.go]: come to think of it I could just use regex

File: engine/resources/exec.go (hunk 17 above)
Comment (adamsigal) [reply engine/resources/exec.go]: Only in the strings package, or should I consider other packages too (like regex)

File: engine/resources/resources_test.go (hunk 18 above)
Comment (adamsigal) [reply engine/resources/resources_test.go]: I don't know what you're referring to

File: engine/resources/exec.go (hunk 13 above)
Comment (purpleidea) [reply engine/resources/exec.go]: It depends on the case. In this case it pulls out the value automatically so you don't need to have it look more complex. Know what I mean?

File: engine/resources/exec.go (hunk 14 above)
Comment (purpleidea) [reply engine/resources/exec.go]: Bingo =D

File: engine/resources/exec.go (hunk 17 above)
Comment (purpleidea) [reply engine/resources/exec.go]: Just look in strings for now. Regex should be used as a last resort.

File: engine/resources/exec.go (hunk 13 above)
Comment (adamsigal) [reply engine/resources/exec.go]: gotcha

File: engine/resources/exec.go (hunk 16 above)
Comment (adamsigal) [reply engine/resources/exec.go]: Am I wrong that it would be more simple to not add a second loop?

File: engine/resources/exec.go (hunk 17 above)
Comment (adamsigal) [reply engine/resources/exec.go]: I can't find a more succinct way in the strings package.

File: engine/resources/exec.go (hunk 16 above)
Comment (purpleidea) [reply engine/resources/exec.go]: There are more readable ways to write this, but after looking at it further, they're less efficient. You can keep it as is, but at least use the loop `range` construct.

File: engine/resources/resources_test.go (hunk 20 above)
Comment (adamsigal) [reply engine/resources/resources_test.go]: I just tested it with a print. It does in fact add the correct variables to env like it is supposed to. However, if I take out necessary code, the test still passes... I'm going to keep working on it.

File: engine/resources/resources_test.go (hunk 18 above)
Comment (adamsigal) [reply engine/resources/resources_test.go]: It's right after the "simple exec" test, it seems well placed to me.

File: engine/resources/resources_test.go (hunk 20 above)
Comment (purpleidea) [reply engine/resources/resources_test.go]: Nowhere in this test are you actually testing the ENV parts :/

File: engine/resources/exec.go (hunk 13 above)
Comment (purpleidea) [reply engine/resources/exec.go:388]: Not sure you did this...

**Comment by purpleidea** [comment]
I think this LGTM!

**Comment by adamsigal** [comment]
> I think this LGTM!

Yes!

**Comment by purpleidea** [comment]
Sorry for the delay in actually merging! It was queued up behind my messing working directory!

All in now \o/
